src/engine/strat_engine/backstore/crypt.rs

```diff
@@ -226,22 +225,73 @@ impl CryptHandle {
         &mut self.device
     }
 
+    /// Create a device handle and load the LUKS2 header into memory from
+    /// a phyiscal path.
+    fn device_from_physical_path(physical_path: &Path) -> Result<Option<CryptDevice>> {
+        let mut device = log_on_failure!(
+            CryptInit::init(physical_path),
+            "Failed to acquire a context for device {}",
+            physical_path.display()
+        );
+        if device
+            .context_handle()
+            .load::<()>(EncryptionFormat::Luks2, None)
+            .is_err()
+        {
+            info!("Device {} is not a LUKS2 device", physical_path.display());
+            Ok(None)
+        } else {
+            Ok(Some(device))
+        }
+    }
+
+    /// Check whether the given physical device can be unlocked with the current
+    /// environment (e.g. the proper key is in the kernel keyring, the device
+    /// is formatted as a LUKS2 device, etc.)
+    // TODO: Use this method when adding blockdevs to a pool to verify that
+    // the same key is being used to encrypt new devices as already encrypted devices.
+    #[allow(dead_code)]
```

Comment:
I _think_ you should be able to test this in a basic way. Its intended use is in the circumstances where a device has already been opened with a given KEK.

So, a valid test would be to:
1. Activate a device. You have now a physical path and a logical path.

3. Test simply by invoking this method on on the physical path you got in step 1. It should return true.

4. Deactivate the device, and test again. I believe this method should still return true on an inactive device.

5. Wipe the device. I believe it should now return false.

Obviously, this is a basic test, since it doesn't check the thing that we'll use it to guard against, which is that the key corresponding to the key description has somehow changed underneath us. Setting that test up would be pretty hard to do at this level in the code.
But the suggested test would exercise the basic functionality, which is still useful.

Reply:
> I _think_ you should be able to test this in a basic way. Its intended use is in the circumstances where a device has already been opened with a given KEK.

That's not entirely correct. I'm happy to update the documentation but this has nothing to do with the device having been activated as much as it checking if the correct key is available and it _could_ be activated. This would work even on a device that has been initialized but not yet activated as long as the correct key is available in the keyring. It runs a test to see if it could be activated with the available keys. Do you still want me to test the same way?

> 
> So, a valid test would be to:
> 
>     1. Activate a device. You have now a physical path and a logical path.
> 
>     2. Test simply by invoking this method on on the physical path you got in step 1. It should return true.
> 
>     3. Deactivate the device, and test again. I believe this method should still return true on an inactive device.
> 
>     4. Wipe the device. I believe it should now return false.
> 
> 
> Obviously, this is a basic test, since it doesn't check the thing that we'll use it to guard against, which is that the key corresponding to the key description has somehow changed underneath us. Setting that test up would be pretty hard to do at this level in the code.
> But the suggested test would exercise the basic functionality, which is still useful.

It will actually test if the key has changed. It is not simply checking for the existence of a key with the correct description but whether or not the given key could actually unlock the device without ever unlocking it.

Reply:
Exactly!

When actually using ```can_unlock``` during initialization, for #1964 it will have to be used on a block device that is already activated, since when adding a device to a pool, it is expected that all the other devices that are worth checking are activated and in use for that pool. So that's probably the most important condition under which to test the function.

Reply:
Ok, now I think I understand a little more now what you're getting at. I think the point of confusion was more that I could see us theoretically using this on a set of closed block devices during device discovery to make sure that all of the devices that are locked can be opened or something similar. This would allow us to check that all are openable before beginning to open some and realizing that others cannot be opened. This may or may not be the way we want to go but it could help from an error reporting standpoint and would also mean that we wouldn't have to roll anything back. For now, I'll test what you've outlined because that will certainly help us either way.

Reply:
Resolved.